onnxruntime/core/providers/coreml/builders/impl/conv_op_builder.cc

```diff
@@ -46,26 +46,68 @@ Status ConvOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, const N
   std::unique_ptr<COREML_SPEC::NeuralNetworkLayer> layer = CreateNNLayer(node);
 
   const auto& input_defs = node.InputDefs();
+  const auto& output_defs = node.OutputDefs();
+  const auto& input_name = input_defs[0]->Name();
+  const auto& output_name = output_defs[0]->Name();
 
   const auto& weight_tensor = *model_builder.GetInitializerTensors().at(input_defs[1]->Name());
   const auto& weight_shape = weight_tensor.dims();
 
+  const bool is_1d_conv = (weight_shape.size() == 3);
+
   NodeAttrHelper helper(node);
-  const auto strides = helper.Get("strides", std::vector<int64_t>{1, 1});
-  const auto onnx_pads = helper.Get("pads", std::vector<int64_t>{0, 0, 0, 0});
-  const auto dilations = helper.Get("dilations", std::vector<int64_t>{1, 1});
+  auto strides = helper.Get("strides", std::vector<int64_t>{1, 1});
+  auto dilations = helper.Get("dilations", std::vector<int64_t>{1, 1});
+  auto onnx_pads = helper.Get("pads", std::vector<int64_t>{0, 0, 0, 0});
+  // Strides/dilations for 1d conv is normally of length 1. Expand them by 1
+  // to meet the required length 2 (for 2d conv it's normally 2)
+  // Similarly 1d conv normally has a length 2 padding. Expand it to length 4 by adding additional zeros.
+  if (is_1d_conv) {
+    if (strides.size() < 2) {
+      strides.push_back(1);
+    }
+    if (dilations.size() < 2) {
+      dilations.push_back(1);
+    }
+    if (onnx_pads.size() < 4) {
+      onnx_pads.insert(onnx_pads.begin() + 1, 0);
+      onnx_pads.push_back(0);
+    }
+  }
   const auto group = helper.Get("group", static_cast<int64_t>(1));
 
   auto* coreml_conv = layer->mutable_convolution();
 
-  coreml_conv->set_outputchannels(weight_shape[0]);  // M
-  coreml_conv->set_kernelchannels(weight_shape[1]);  // C/Group
-  coreml_conv->add_kernelsize(weight_shape[2]);      // H
-  coreml_conv->add_kernelsize(weight_shape[3]);      // W
+  std::string expand_output_name = model_builder.GetUniqueName(node.Name() + "_expandDims");
+  auto weight_tensor_prime = weight_tensor;
+
```

Comment:
Can we make weight_tensor a copy early on (instead of the delayed copy for weight_tensor_prime) so that we can make weight_shape mutable and add 1 to the dims of that if it's a 1D conv early on?

Possibly that lets you just conditionally add the expand layer, but use common code for the coreml_conv layer and HandleAutoPad logic so the binary size cost is smaller.

Reply:
Thanks for the suggestion. That's right! Updated.